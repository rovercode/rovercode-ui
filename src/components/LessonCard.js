import React from 'react';
import { hot } from 'react-hot-loader';
import {
  Card,
  CardHeader,
  CardActionArea,
  Grid,
  Link,
  SvgIcon,
  Typography,
} from '@material-ui/core';
import { CheckCircle, Lens, Lock } from '@material-ui/icons';
import { makeStyles } from '@material-ui/core/styles';
import { grey } from '@material-ui/core/colors';
import { FormattedMessage } from 'react-intl';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
  inProgress: {
    color: theme.palette.primary.main,
  },
  complete: {
    color: theme.palette.success.main,
  },
  unavailable: {
    color: grey[500],
  },
}));

const LessonCard = ({ lesson, userTier, onClick }) => {
  const classes = useStyles();

  const { progress } = lesson.state || {};
  const unavailable = progress === 'UNAVAILABLE';

  const progressIcon = () => {
    switch (progress) {
      case 'IN_PROGRESS':
        return (<Lens className={classes.inProgress} />);

      case 'COMPLETE':
        return (<CheckCircle className={classes.complete} />);

      default:
        return (null);
    }
  };

  const subheader = () => (
    userTier >= lesson.tier ? (
      lesson.description
    ) : (
      <Grid container direction="row">
        <Grid item xs={2}>
          <Lock fontSize="large" />
        </Grid>
        <Grid item xs={10}>
          <Typography variant="subtitle1">
            <Link href="/user/settings" color="textSecondary" underline="always">
              <FormattedMessage
                id="app.lesson_card.blocked1"
                description="Informs the user that upgrade is required to access"
                defaultMessage="Upgrade"
              />
            </Link>
            {' '}
            <FormattedMessage
              id="app.lesson_card.blocked2"
              description="Informs the user that upgrade is required to access"
              defaultMessage="your account to access this lesson."
            />
          </Typography>
        </Grid>
      </Grid>
    )
  );

  const handleOnClick = (e) => {
    if (userTier >= lesson.tier) {
      onClick(e);
    }
  };

  return (
    <Card className={unavailable ? classes.unavailable : ''}>
      <CardActionArea
        id={lesson.active_bd}
        data-owned={lesson.active_bd_owned}
        onClick={handleOnClick}
        disabled={unavailable}
      >
        <CardHeader
          title={lesson.reference}
          subheader={subheader()}
          action={(
            <SvgIcon>
              {progressIcon()}
            </SvgIcon>
          )}
        />
      </CardActionArea>
    </Card>
  );
};

LessonCard.defaultProps = {
  userTier: 1,
};

LessonCard.propTypes = {
  lesson: PropTypes.shape({
    id: PropTypes.number.isRequired,
    reference: PropTypes.string.isRequired,
    description: PropTypes.string,
    tier: PropTypes.number,
    active_bd: PropTypes.number.isRequired,
    active_bd_owned: PropTypes.bool.isRequired,
    state: PropTypes.shape({
      progress: PropTypes.string,
    }),
  }).isRequired,
  userTier: PropTypes.number,
  onClick: PropTypes.func.isRequired,
};

export default hot(module)(LessonCard);
