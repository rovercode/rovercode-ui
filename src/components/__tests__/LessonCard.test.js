import React from 'react';
import { shallow } from 'enzyme';
import { CardActionArea, CardHeader } from '@material-ui/core';

import LessonCard from '../LessonCard';

describe('The LessonCard component', () => {
  const lesson = {
    id: 1,
    reference: 'Test',
    description: 'This is a test',
    active_bd: 1,
    active_bd_owned: false,
    tier: 1,
  };
  const onClick = jest.fn();
  beforeEach(() => {
    onClick.mockReset();
  });

  test('renders on the page with no errors', () => {
    lesson.state = { progress: 'AVAILABLE' };
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(CardActionArea).prop('disabled')).toBe(false);
    expect(wrapper.find(CardHeader).prop('action').props.children).toBeNull();
  });

  test('shows progress icon', () => {
    lesson.state = { progress: 'IN_PROGRESS' };
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} />);

    expect(wrapper.find(CardActionArea).prop('disabled')).toBe(false);
    expect(wrapper.find(CardHeader).prop('action').props.children.props.className).toBe('makeStyles-inProgress-1');
    expect(wrapper.find(CardHeader).prop('action').props.children.type.displayName).toBe('LensIcon');
  });

  test('shows complete icon', () => {
    lesson.state = { progress: 'COMPLETE' };
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} />);

    expect(wrapper.find(CardActionArea).prop('disabled')).toBe(false);
    expect(wrapper.find(CardHeader).prop('action').props.children.props.className).toBe('makeStyles-complete-2');
    expect(wrapper.find(CardHeader).prop('action').props.children.type.displayName).toBe('CheckCircleIcon');
  });

  test('is disabled when unavailable', () => {
    lesson.state = { progress: 'UNAVAILABLE' };
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} />);

    expect(wrapper.find(CardActionArea).prop('disabled')).toBe(true);
    expect(wrapper.find(CardHeader).prop('action').props.children).toBeNull();
  });

  test('handles missing progress', () => {
    lesson.state = null;
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} />);

    expect(wrapper.find(CardActionArea).prop('disabled')).toBe(false);
    expect(wrapper.find(CardHeader).prop('action').props.children).toBeNull();
  });

  test('allows click when unlocked', () => {
    lesson.state = { progress: 'IN_PROGRESS' };
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} />);

    wrapper.find(CardActionArea).simulate('click');
    expect(onClick).toHaveBeenCalled();
  });

  test('blocks click when locked', () => {
    lesson.state = { progress: 'IN_PROGRESS' };
    const wrapper = shallow(<LessonCard lesson={lesson} onClick={onClick} userTier={0} />);

    wrapper.find(CardActionArea).simulate('click');
    expect(onClick).not.toHaveBeenCalled();
  });
});
