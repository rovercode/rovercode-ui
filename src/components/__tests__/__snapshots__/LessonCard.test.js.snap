// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The LessonCard component renders on the page with no errors 1`] = `
<WithStyles(ForwardRef(Card))
  className=""
>
  <WithStyles(ForwardRef(CardActionArea))
    data-owned={false}
    disabled={false}
    id={1}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(CardHeader))
      action={<WithStyles(ForwardRef(SvgIcon)) />}
      subheader="This is a test"
      title="Test"
    />
  </WithStyles(ForwardRef(CardActionArea))>
</WithStyles(ForwardRef(Card))>
`;
